tests/drivers/i2c/i2c_emul/src/test_forwarding.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <functional>
+
+#define CUSTOM_FFF_FUNCTION_TEMPLATE(RETURN, FUNCNAME, ...)                                        \
+	std::function<RETURN(__VA_ARGS__)> FUNCNAME
+
+#include "emulated_target.h"
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/fff.h>
+#include <zephyr/ztest.h>
+
+DEFINE_FFF_GLOBALS;
+
+#define CONTROLLER_LABEL DT_NODELABEL(i2c0)
+#define TARGET_LABEL     DT_NODELABEL(i2c1)
+
+namespace
+{
+/* Get the devicetree constants */
+constexpr const struct device *controller = DEVICE_DT_GET(CONTROLLER_LABEL);
+constexpr const struct device *target = DEVICE_DT_GET(TARGET_LABEL);
+constexpr const uint16_t forward_address = DT_PHA_BY_IDX(CONTROLLER_LABEL, forwards, 0, addr);
+
+BUILD_ASSERT(DEVICE_DT_GET(DT_PHANDLE_BY_IDX(CONTROLLER_LABEL, forwards, 0)) == target,
+	     "Forwarding misconfigured");
+
+/* Setup the fake functions for the I2C target. */
+FAKE_VALUE_FUNC(int, target_read_requested, struct i2c_target_config *, uint8_t *);
+FAKE_VALUE_FUNC(int, target_read_processed, struct i2c_target_config *, uint8_t *);
+FAKE_VALUE_FUNC(int, target_write_requested, struct i2c_target_config *);
+FAKE_VALUE_FUNC(int, target_write_received, struct i2c_target_config *, uint8_t);
+FAKE_VALUE_FUNC(int, target_stop, struct i2c_target_config *);
+
```

Comment:
Can this move into the emulated_target.c file?
Or maybe just move the contents of emulated_target.c here.

Reply:
Sure, I had to turn emulated_target to a CPP file since the fakes are using lambdas, but it's much cleaner now.